core_backend/app/llm_call/parse_input.py

```diff
@@ -2,75 +2,196 @@
     """
-    return getattr(
-        SafetyClassification,
-        _ask_llm(query_text, SafetyClassification.get_prompt()),
-    )
+    if response.state != ResultState.ERROR:
+        safety_classification = getattr(
+            SafetyClassification,
+            _ask_llm(question.query_text, SafetyClassification.get_prompt()),
+        )
+        if safety_classification != SafetyClassification.SAFE:
+            response.llm_response = (
+                "Sorry, we are unable to answer your question."
+                "Please rephrase your question and try again."
+            )
+            response.state = ResultState.ERROR
+        response.debug_info["safety_classification"] = safety_classification.value
+
+    return question, response
 
 
-def input_is_safe(question: UserQueryRefined) -> bool:
+def identify_language(func: Callable) -> Callable:
     """
-    Checks for prompt injection and inappropriate language in the question.
+    Decorator to identify the language of the question.
     """
-    if classify_safety(question.query_text) == SafetyClassification.SAFE:
-        return True
-    else:
-        return False
-
 
-def identify_language(question: UserQueryRefined) -> UserQueryRefined:
+    @wraps(func)
+    async def wrapper(
+        question: UserQueryRefined,
+        response: UserQueryResponse,
+        *args: Any,
+        **kwargs: Any,
+    ) -> UserQueryResponse:
+        """
+        Wrapper function to identify the language of the question.
+        """
+        question, response = _identify_language(question, response)
+        response = await func(question, response, *args, **kwargs)
+        return response
+
+    return wrapper
+
+
+def _identify_language(
+    question: UserQueryRefined, response: UserQueryResponse
+) -> Tuple[UserQueryRefined, UserQueryResponse]:
     """
     Identifies the language of the question.
     """
+    if response.state != ResultState.ERROR:
+        question.original_language = getattr(
+            IdentifiedLanguage,
+            _ask_llm(question.query_text, IdentifiedLanguage.get_prompt()),
+        )
+        if question.original_language is not None:
+            response.debug_info["original_language"] = question.original_language.value
 
-    question.original_language = getattr(
-        IdentifiedLanguage,
-        _ask_llm(question.query_text, IdentifiedLanguage.get_prompt()),
-    )
+    return question, response
 
-    return question
 
+def translate_question(func: Callable) -> Callable:
+    """
+    Decorator to translate the question.
+    """
 
-def translate_question(question: UserQueryRefined) -> UserQueryRefined:
+    @wraps(func)
+    async def wrapper(
+        question: UserQueryRefined,
+        response: UserQueryResponse,
+        *args: Any,
+        **kwargs: Any,
+    ) -> UserQueryResponse:
+        """
+        Wrapper function to translate the question.
+        """
+        question, response = _identify_language(question, response)
```

Comment:
I made the decision to run `identify language` everytime `translate question` is called. But there is no need to do that. Se line 141, we can just require the user to run `identify language` and throw an exception if they haven't.